primitives/src/vstaging/executor_params.rs

```diff
@@ -0,0 +1,123 @@
+// Copyright 2021 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Abstract execution environment parameter set.
+//!
+//! Parameter set is encoded as an opaque vector which structure depends on the execution
+//! environment itself (except for environment type/version which is always represented
+//! by the first element of the vector). Decoding to a usable semantics structure is
+//! done in `polkadot-node-core-pvf`.
+
+use parity_scale_codec::{Decode, Encode};
+use parity_util_mem::MallocSizeOf;
+use scale_info::TypeInfo;
+use sp_std::{ops::Deref, vec::Vec};
+
+/// Execution environment type
+#[derive(Clone, Copy, Debug, Encode, Decode, PartialEq, Eq, MallocSizeOf, TypeInfo)]
+pub enum ExecutionEnvironment {
+	/// Generic Wasmtime executor
+	WasmtimeGeneric = 0,
+}
+
+/// Executor instantiation strategy
+#[derive(Clone, Debug, Encode, Decode, PartialEq, Eq, MallocSizeOf, TypeInfo)]
+pub enum ExecInstantiationStrategy {
+	/// Pooling copy-on-write
+	PoolingCoW,
+	/// Recreate instance copy-on-write
+	RecreateCoW,
+	/// Pooling
+	Pooling,
+	/// Recreate instance
+	Recreate,
+	/// Legacy instantiation strategy
+	Legacy,
+}
+
+/// A single executor parameter
+#[non_exhaustive]
+#[derive(Clone, Debug, Encode, Decode, PartialEq, Eq, MallocSizeOf, TypeInfo)]
+pub enum ExecutorParam {
+	/// General parameters:
+	/// Execution environment type and a version of the set of execution parameters, unique in
+	/// the scope of execution environment type
+	Version(ExecutionEnvironment, u32),
```

Comment:
IIRC, we agreed that we don't need a generic version. But once again, that depends on the design of this feature. Here is a recap of the design space. 

Instead of those configuration parameters, theoretically, we could get away with a single number. That number represents a version of the execution environment.

Let's imagine we want to gracefully increase the logical stack limit. Assuming that the current version is 0, the version with the increased logical stack limit will be assigned to 1. The runtime API would return the version of EE as a number, trickling into the PVF host via the candidate validation through backing, approvals, etc. The PVF host would configure the stack limit depending on the version. The nodes that were not updated to version 1 would refrain from participating in the parachain consensus and notify their operators.

However, while it works, this is a heavy-weight approach. If we only want to tune one small parameter, we would require to release a new version of the node, distribute it among the validators, wait until they are updated, and so on. 

The advantage of the approach with individual parameters is that already present parameter knobs do not require updating the node.

Reply:
We do need to communicate the changes happening upstream, that is Substrate Runtime environment. 

For example, if `sc_executor_wasmtime` added some new wasm feature, it would have to be enabled consistently across the validator population. But that's a boolean, not a version.

Maybe I am missing something?